core/tests/mod.rs

```diff
@@ -0,0 +1,646 @@
+                                Message::BlockSyncUpdate(block.clone().into()),
+                            ));
+                        }
+                    }
+                    last_block_sync = Instant::now();
+                }
+
+                {
+                    let sumeragis = sumeragis.lock().unwrap();
+                    let mut correct_count = 0;
+                    for sumeragi in sumeragis.values() {
+                        if sumeragi.apply_wsv(|wsv| {
+                            wsv.block_hashes_after_hash(None).len() == block_height
+                        }) {
+                            correct_count += 1;
+                        }
+                    }
+                    if correct_count == peer_ids.len() {
+                        success = true;
+                        break;
+                    }
+                }
+                sleep_ms(100);
+            }
+            assert!(success, "Not all sumeragi instances committed the block.");
+        }
+
+        *should_softfork.lock().unwrap() = true;
+
+        let account_id: AccountId = "alice@wonderland".parse().unwrap();
+        let asset_definition_id: AssetDefinitionId = "rose#wonderland".parse().expect("Valid");
+        let mint_asset = MintBox::new(
+            1_u32.to_value(),
+            IdBox::AssetId(AssetId::new(
+                asset_definition_id.clone(),
+                account_id.clone(),
+            )),
+        );
+
+        let transaction =
+            TransactionBuilder::new(account_id.clone(), vec![mint_asset.into()], 100_000)
+                .sign(get_key_pair())
+                .unwrap();
+        let transaction = AcceptedTransaction::accept::<false>(
+            transaction,
+            &blank_config().wsv.transaction_limits,
+        )
+        .unwrap();
+
+        for peer_id in &peer_ids {
+            let sumeragis = sumeragis.lock().unwrap();
+            let wsv = sumeragis
+                .get(peer_id)
+                .unwrap()
+                .apply_wsv(WorldStateView::clone);
+            let queue = queues.get(peer_id).unwrap();
+            queue.push(transaction.clone().into(), &wsv).unwrap();
+        }
+    }
+}
+
+#[allow(clippy::type_complexity)]
+fn set_up_network(
+    peer_count: usize,
+    post_proc: Arc<
+        (dyn Fn(MessagePacket, &PeerId, &Arc<Mutex<HashMap<PeerId, SumeragiHandle>>>)
+             + Send
+             + Sync),
+    >,
+) -> (
+    Vec<KeyPair>,
+    Vec<PeerId>,
+    Vec<Configuration>,
+    HashMap<PeerId, Arc<Queue>>,
+    Arc<Mutex<HashMap<PeerId, SumeragiHandle>>>,
+) {
+    set_up_network_with_genesis(peer_count, true, post_proc)
+}
+
+#[allow(clippy::needless_pass_by_value, clippy::type_complexity)]
+fn set_up_network_with_genesis(
+    peer_count: usize,
+    commit_genesis: bool,
+    post_proc: Arc<
+        (dyn Fn(MessagePacket, &PeerId, &Arc<Mutex<HashMap<PeerId, SumeragiHandle>>>)
+             + Send
+             + Sync),
```

Comment:
No. This is not how you do it. 

1. You should not have an atomically reference counted function pointer. EVER. The functional approach is to take this by reference. 
2. I don't think a function pointer is a good choice here. See https://stackoverflow.com/questions/56420597/should-i-pass-function-objects-by-value-or-by-reference. You are actively making the compiler do more work. The end result is very likely the same thanks to LLVM's devirtualiser. 
3. And this more contentious; the type that holds this function value, should itself be generic. It can make things a bit uglier, and before GATs far uglier than they would now. Still that's more clear with the intent. 
4. This should really be a `struct -> struct` function or better yet `impl From<ArgStruct> for ReturnStruct`. Rust has robust destructuring, the only way a tuple is different from a struct is in that it's ambiguous about names, less verbose and far less error-prone.